### ISSUE TITLE
Amplify multi-env setup causing my production build with wrong env access

### ISSUE THREAD

**Issue by hsiehro** [aws-amplify/amplify-cli#2182]
Hi, 

Firstable, I know this might not be the proper place to raise Amplify multi-environment issue. But since I'm not getting feedback on AWS forum, I hope to address the issue here to get some attention from AWS team.

I'm using amplify react native to build both ios and Android application. I created two environment branch on both git and amplify - development and production. Recently I built my application and ran into a very strange problem. 

1. Login in user (cognito from production env) 
2. Query data on my app ( api graphql from production env)
3. Force close the app or logout and login again (the new login cognito user switch to dev environment - this registered username exist on both dev and prod env, that's why it can login without problem, otherwise it will fail to login )
4. Query data (api graphql switch to dev env)

As you can see from above, all the sudden my whole app switch to development environment even though the build is intend for production. How is it possible to have mix env co-exist in my app? The way I handle env build is like this:

When working on dev environment:
1. git commit to dev
2. git push to dev
3. amplify init or env checkout to dev
4. amplify push (if any changes)
5. build the dev image

When switch to production environment:
1. git checkout master which is production env
2. git merge dev to master which is production env
3. git push to master which is production
4. amplify env checkout to production env
5. amplify push (if any changes)
6. build the production image

The way I make sure my amplify environment setup switch to production is by checking inside the aws-exports.js file, there are no dev config in that file. And as I stated above, my first launch always points me correctly to the production server. Then how is it possible with some app action change resulting from a production switch to development server? Is there any place I missed causing such sudden switch?

I really hope to get feedback from the AWS Amplify development team. This issue is preventing me from launching my app to the store. It would be a disaster if I didn't run enough testing before launch with such a major defect.

In addition, I first suspect the config might be cache somewhere in the device during my development testing, but tried a brand new device and the problem persist. So that is not the case. 

Thanks in advance for the support.

**Comment by hsiehro** [comment]
Several days with no feedback, it makes me worried whether or not Amplify itself is still a highly active maintain project inside Amazon. If any plan to phase out in the next year or two, please let me know. Thanks.

**Comment by kaustavghosh06** [comment]
@hsiehro Which switching environments are you sure that the `aws_exports.js` generated by the CLI is different and based on your environment?
I believe it is required for you to restart your react native app after you change environments (which changes your aws_exports.js file), for the app to pick up the changes - if you don't have any watcher or hot-reload setup.
@undefobj @mlabieniec Any thoughts?

**Comment by hsiehro** [comment]
@kaustavghosh06 Hi, appreciated your prompt feedback. Yes, everytime I made the env switch, I always double confirm to make sure the config inside aws-exports.js is correctly setup before I make the build. I assume that is the only place I could check.

I use expo to develop and test my application before each build, so the expo itself already have a hot-reload mechanism inside. The obvious problem is when I first open the application, I can tell I was on the production env by rendering the page. But somehow if I logout and force close and re-launch my app again, all the sudden page rendering shows me I was in development environment with all the test data inside. Hope this can give some clue...Thanks~

**Comment by kaustavghosh06** [comment]
Contacted @haverchuck to take a look if the Amplify JS lib picks up configurations from a cache on the device (Which looks like its stale in your case). We're investigating and trying to reproduce it.

**Comment by haverchuck** [comment]
@hsiehro - Are you using the [Amplify GraphQL Client](https://aws-amplify.github.io/docs/js/api) or the [AWS AppSync SDK](https://aws-amplify.github.io/docs/js/api#aws-appsync-sdk)?

**Comment by hsiehro** [comment]
@kaustavghosh06 @haverchuck Thanks for the help. For your information, I tested on three devices:
iPhone 6 - main development and test device for ios
Oppo - main development and test device for Android
iPhone 6+ - a friend's device only use the final ipa production build I generated to run the test (clean device and 1st install, no previous cache data from any of the build)

I original thought also assume all the config were cache on the device somehow, but use a new clean device still had the problem proves me wrong.

@haverchuck I assume I'm using AppSync as part of aws-amplify sdk. This is what is shown inside my aws-exports file - "aws_appsync_graphqlEndpoint": "https://xxxx

Thanks again for looking into this issue.

**Comment by haverchuck** [comment]
@hsiehro -  The AppSync SDK is a seperate npm package that can be used with or without Amplify.   The Amplify library has it's own seperate GraphQL client, which is lighter weight.  
For your information the distinction can be found [here](https://aws-amplify.github.io/docs/js/api#using-aws-appsync).  You don't necessarily need it to use Amplify with AppSync, though.  It sounds like you are using Amplify without importing the AppSync SDK.

This is relevant because the AppSync SDK gives you the ability to set a[ fetch policy,](https://aws-amplify.github.io/docs/js/api#run-a-query) which influences how the cache is used. 

I'm continuing to try to reproduce.

**Comment by hsiehro** [comment]
@haverchuck Thanks for the explanation. In that case, I am using the GraphQL client without installing AppSync SDK separately, which you are correct.

**Comment by undefobj** [comment]
@hsiehro in your app can you log out the contents of your aws-exports at runtime to see which environment config is currently being used?

**Comment by hsiehro** [comment]
@undefobj  Not sure what you mean by log out at runtime. When first launch the application, it always direct user to a page where he/she can login/logout and change password...etc. If you need, I can send you the app depends on ios or Android device you are using. Thanks.

**Comment by undefobj** [comment]
Something like `console.log(aws-exports)` with remote debugging to see what the values actually being used are as it sounds like you have some race condition/conflict with the env being used.

**Comment by hsiehro** [comment]
@undefobj I see, I also wish to add log message to it to trace where it gets its value from, but my question still is if there is no dev environment config files existed inside the build, how can it obtain all the right paths, keys and values to access different server?

**Comment by hsiehro** [comment]
@undefobj Screenshot of my current source folder structure. Let me know if you need to look at particular file. Thanks.

![image](https://user-images.githubusercontent.com/8207338/64227711-0cf39c00-cf17-11e9-8b06-0267287262f7.png)

**Comment by haverchuck** [comment]
@hsiehro -  You can see the what is actually in the app build at runtime using Expo's[ remote debugging option](https://docs.expo.io/versions/latest/workflow/debugging/#on-ios-simulator).

**Comment by hsiehro** [comment]
@haverchuck Thanks for info. One thing I didn't mention is I was never able to reproduce it when running on expo. The reason is because expo runs by hot-reload, it doesn't give me the option of exit the app or force close and re-enter again. The way to 100% reproduce this issue is to force close my app and re-launch. Another problem is the logout page only shows at startup, once entering application main screen, i have no method to logout. I guess I could add a button to run some test on it. If really need to check from runtime, what i can do is to run it on Android's logcat to see what is going on. Let me give it a try and get back to you. Thanks.

**Comment by haverchuck** [comment]
@hsiehro You are running the app outside of expo to repro this?  Did you eject it?

**Comment by hsiehro** [comment]
@haverchuck This is my current situation. I develop and test my application between dev and production env using expo because of its hot-reload feature, I can immediate see the result.  Everything is working as expected. Then one day I figure I am ready to make a standalone build for both ios and android, so I used expo build command to create apk and ipa files and installed on my device.

Initial test, everything was fine. I can see data from production env when app launch, and all feature are working as expected. It's only because I never quit the application. As what most mobile user will do, for iOS, you press the home buttom to exit from App. Android, one quick way is to go the the application list button and swipe or close it to prevent it running in the background. Then next time open the application by pressing the icon, then the whole env switched to dev server. Yes, you can call such eject as normal user behavior.

**Comment by hsiehro** [comment]
@haverchuck Correction on the iOS part, double press the home key and swipe to kill the application in the background. Thanks.

**Comment by hsiehro** [comment]
So far, I couldn't find information from logcat as what makes the switch. Is there a way to trace what's happening inside Auth.configure(awsconfig)? My awsconfig is import from aws-exports, and there is only one file exist in that path. This is the only place I could think of causing the environment change, as part of the initialization. And further thinking, this could also be an Expo problem too since we don't know what happen during the build. I will do more testing. Thanks.

By the way, I know it's kind of late for you guys. Let's stop for today and many thanks for helping out.

**Comment by hsiehro** [comment]
@haverchuck Sorry I misunderstood your original question. I am using pure expo during development and did not eject from it. The reason is because expo help me build an ios image without a macbook.

**Comment by haverchuck** [comment]
@hsiehro If you use remote debugging with expo on an emulator, I believe you can take full advantage of Chrome dev tools.  So you should be able to put a breakpoint inside of Auth.configure.  You can also try turning on Amplify's built-in [Logger](https://aws-amplify.github.io/docs/js/logger) utility.

**Comment by hsiehro** [comment]
@haverchuck Thanks for the advice. First to jump into the conclusion. This problem has nothing to do with the Amplify env itself, my apology for all of your time spending. It is caused by the previous setting that was cached and somehow mixup during the expo build. The way I come to this conclusion is to do an expo start -c which -c flag is to clear cache first before creating a new build. The problem seems to disappeared, and I will run more testing to confirm this result.

Yesterday I already suspected the problem is more likely within the expo itself. Because after trace the code, the only place for amplify config init call in my application is by calling Auth.configure, and it is pointing to the right exports file, while the the switch occur during runtime.

Yes, I did take the advantage of using expo debug and tried to setup breakpoint and print logs at the same time, but not much of help. This problem is really not within the application logic itself, it's more like Auth.config is getting its new value somewhere at certain executing state. I can only see the switch happen but not able to monitor what happen in between.

At last, sorry again for wasting much of your time. I really appreciated all of your great support. Keep up the good work. Amplify is something I would highly recommended for everyone, it really saves your development time. Thank you.

After do more testing and verify my result, I will go ahead and close this issue.
